File: javascript/phone-number/phone-number.js
export const clean = (number) => {
  let clean_num = ""

  for (const part of number) {
    if (['(', ')', '-', '+', '.', ' '].includes(part)) {
      continue
    }

    if (part.match(/[a-z]/i)) {
      throw new Error('Letters not permitted')
    }

    if (isNaN(Number(part))) {
      throw new Error('Punctuations not permitted')
    }

    clean_num += part
  }

  if (clean_num.length < 10) {
    throw new Error('Incorrect number of digits')
  }

  if (clean_num.length > 11) {
    throw new Error('More than 11 digits')
  }

  if (clean_num.length == 11) {
    if (clean_num[0] !== '1') {
      throw new Error('11 digits must start with 1')
    }

    clean_num = clean_num.substring(1);
  }

  switch(clean_num[0]) {
    case '0':
      throw new Error('Area code cannot start with zero')
    case '1':
      throw new Error('Area code cannot start with one')
  }

  switch(clean_num[3]) {
    case '0':
      throw new Error('Exchange code cannot start with zero')
    case '1':
      throw new Error('Exchange code cannot start with one')
  }

  return clean_num
}

File: python/phone-number/phone_number.py
import string

class PhoneNumber:
    def __init__(self, number):
        number_count = 0
        sanitized_number = ""
        pos = 0

        for n in number:
            if n.isnumeric():
                number_count += 1

            if n.isalpha():
                raise ValueError('letters not permitted')

            if n not in ['(', ')', '.', '-', '+'] and n in list(string.punctuation):
                raise ValueError('punctuations not permitted')

        if number_count > 11:
            raise ValueError('must not be greater than 11 digits')
        if number_count < 10:
            raise ValueError('must not be fewer than 10 digits')

        for n in number[::-1]:
            if pos == 6:
                if n == '0':
                    raise ValueError('exchange code cannot start with zero')
                if n == '1':
                    raise ValueError('exchange code cannot start with one')
            if pos == 9:
                if n == '0':
                    raise ValueError('area code cannot start with zero')
                if n == '1':
                    raise ValueError('area code cannot start with one')

            if n.isnumeric():
                sanitized_number += n
                pos += 1

        sanitized_number = sanitized_number[::-1]

        if pos == 11:
            country = sanitized_number[0]
            sanitized_number = sanitized_number[1:]
            if country != '1':
                raise ValueError('11 digits must start with 1')

        self.number = sanitized_number
        self.area_code = sanitized_number[:3]
        self.exchange = sanitized_number[3:6]
        self.rest = sanitized_number[6:]

    def pretty(self):
        return f"({self.area_code})-{self.exchange}-{self.rest}"
        
